toolchain/sem_ir/clang_source_location.h

```diff
@@ -0,0 +1,22 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEM_IR_CLANG_SOURCE_LOCATION_H_
+#define CARBON_TOOLCHAIN_SEM_IR_CLANG_SOURCE_LOCATION_H_
+
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Basic/SourceLocation.h"
+
+namespace Carbon::SemIR {
+
+// Contains the Clang source location and a pointer to the diagnostics engine so
+// it can be transformed to a path and line number.
+struct ClangSourceLocation {
+  clang::SourceLocation source_location;
+  const clang::DiagnosticsEngine* diag_engine;
```

Comment:
Why does each location need its own diagnostic engine? That is, why isn't this per-file instead of per-location?

Reply:
I see also this is stored as part of ASTUnit, so maybe we don't need additional pointers to it?

Reply:
There were two reasons I opted for that:
1. It avoids the implicit assumption that the same `DiagnosticsEngine` is used in both places (though there's also the question of ownership).
2. We don't have the `ASTUnit` in the context when `HandleDiagnostic()` is called since it's being called while the AST is generated.
I've changed that and replaced `ClangSourceLocation` with `clang::SourceLocation`, by delaying emitting the diagnostics after `buildASTFromCodeWithArgs()` is finished.

Reply:
We care about keeping these structures small, and minimizing RAM consumption. As a consequence, I think we need to not store the same pointer per location if there's a way to avoid it.

Reply:
Thanks!
Makes sense.